SYCL/ESIMD/api/functional/value_conv.hpp

```diff
@@ -0,0 +1,117 @@
+
+#pragma once
+
+#include "value.hpp"
+
+namespace esimd_test::api::functional {
+
+// Utility class to retrieve specific values for tests depending on the source
+// and destination data types. May be used to retrieve converted reference data.
+// All provided methods are save to use and protected from UB when call
+// static_cast<int>(unsigned int).
+template <typename SrcT, typename DstT> struct value_conv {
+  static SrcT lowest() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::lowest());
+      } else {
+        return value<SrcT>::lowest();
+      }
+    } else {
+      return std::max(value<SrcT>::lowest(),
+                      static_cast<SrcT>(value<DstT>::lowest()));
+    }
+  }
+
+  static SrcT max() {
+    if constexpr (!sts::is_unsigned_v<SrcT> && sts::is_unsigned_v<DstT>) {
+      if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+        return static_cast<SrcT>(value<DstT>::max());
+      } else {
+        return value<SrcT>::max();
+      }
+    } else {
+      return std::min(value<SrcT>::max(),
+                      static_cast<SrcT>(value<DstT>::max()));
+    }
+  }
+
+  static SrcT denorm_min() {
+    if constexpr (std::is_same_v<SrcT, sycl::half> ||
+                  std::is_same_v<DstT, sycl::half>) {
+      return value<SrcT>::denorm_min();
+    } else if constexpr (sizeof(SrcT) > sizeof(DstT)) {
+      return static_cast<SrcT>(value<DstT>::max());
+    } else {
+      return value<SrcT>::max();
+    }
+  }
+};
+
+// Provides std::vector with the reference data according to the obtained data
+// types and number of elements.
+template <typename SrcT, typename DstT, int NumElems>
+std::vector<SrcT> generate_ref_conv_data() {
+  static_assert(std::is_integral_v<SrcT> ||
+                    type_traits::is_sycl_floating_point_v<SrcT>,
+                "Invalid source type.");
+  static_assert(std::is_integral_v<DstT> ||
+                    type_traits::is_sycl_floating_point_v<DstT>,
+                "Invalid destination type.");
+
+  static const SrcT min = value_conv<SrcT>::lowest();
+  static const SrcT min_half = min / 2;
+  static const SrcT max = value_conv<SrcT>::max();
+  static const SrcT max_half = max / 2;
+  static const SrcT nan = value<SrcT>::nan();
+  static const SrcT inf = value<SrcT>::inf();
+  static const SrcT denorm = value_conv<SrcT, DstT>::denorm_min();
+  static const SrcT positive = static_cast<SrcT>(126.75);
+
+  std::vector<SrcT> ref_data;
+
+  if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                type_traits::is_sycl_floating_point_v<DstT>) {
+    ref_data = details::construct_ref_data<SrcT, NumElems>(
+        {min, max, -0.0, +0.0, 0.1, denorm, nan, -inf});
+  } else if constexpr (type_traits::is_sycl_floating_point_v<SrcT> &&
+                       std::is_unsigned_v<DstT>) {
```

Comment:
I suggest do it in separate PR because we need more time to avoid UB cases
I leave the TODO comment [here](https://github.com/intel/llvm-test-suite/pull/696/commits/5959fb1f2f4dd5aca794998f3a625824bfc42107#diff-e031be0968a4e0cad81b5b2879e34920db09777fa3165f4f722fb4e0ab560826R52-R53)